File: solutions/2026-03-05/java/2026-03-05-minimum-changes-to-make-alternating-binary-string.java
```java
/**
 * Problem: Minimum Changes To Make Alternating Binary String
 * Problem Summary: Given a binary string, find the minimum operations to make it alternating
 * (no adjacent characters are the same). An operation consists of changing '0' to '1' or '1' to '0'.
 * Link: https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string/
 *
 * Approach:
 * An alternating binary string can start with either '0' or '1'.
 * There are only two possible target alternating strings:
 * 1. Starts with '0': "010101..."
 * 2. Starts with '1': "101010..."
 *
 * We can calculate the number of changes required to transform the input string `s` into each of these two target strings.
 * The minimum of these two counts will be our answer.
 *
 * To do this efficiently, we can iterate through the input string `s` once.
 * For each character `s[i]` at index `i`:
 * - If the target string starts with '0', the character at index `i` should be '0' if `i` is even, and '1' if `i` is odd.
 * - If the target string starts with '1', the character at index `i` should be '1' if `i` is even, and '0' if `i` is odd.
 *
 * We maintain two counters: `changes_for_start_0` and `changes_for_start_1`.
 * For each character `s[i]`:
 * - If `i` is even:
 *   - If `s[i]` is '1', it doesn't match the '0101...' pattern, so increment `changes_for_start_0`.
 *   - If `s[i]` is '0', it doesn't match the '1010...' pattern, so increment `changes_for_start_1`.
 * - If `i` is odd:
 *   - If `s[i]` is '0', it doesn't match the '0101...' pattern, so increment `changes_for_start_0`.
 *   - If `s[i]` is '1', it doesn't match the '1010...' pattern, so increment `changes_for_start_1`.
 *
 * Finally, return `min(changes_for_start_0, changes_for_start_1)`.
 *
 * Time Complexity: O(N), where N is the length of the string `s`. We iterate through the string once.
 * Space Complexity: O(1), as we only use a few integer variables for counters.
 */
class Solution {
    /**
     * Calculates the minimum number of operations to make a binary string alternating.
     *
     * @param s The input binary string.
     * @return The minimum number of operations.
     */
    public int minOperations(String s) {
        // Counter for changes needed if the target alternating string starts with '0' (e.g., "0101...")
        int changesForStart0 = 0;
        // Counter for changes needed if the target alternating string starts with '1' (e.g., "1010...")
        int changesForStart1 = 0;

        // Iterate through the input string `s`
        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);

            // Check for the alternating pattern starting with '0'
            // If index `i` is even, the character should be '0'.
            // If index `i` is odd, the character should be '1'.
            if (i % 2 == 0) { // Even index, expected '0' for "0101..." pattern
                if (currentChar == '1') {
                    changesForStart0++; // Mismatch, needs a change
                }
            } else { // Odd index, expected '1' for "0101..." pattern
                if (currentChar == '0') {
                    changesForStart0++; // Mismatch, needs a change
                }
            }

            // Check for the alternating pattern starting with '1'
            // If index `i` is even, the character should be '1'.
            // If index `i` is odd, the character should be '0'.
            if (i % 2 == 0) { // Even index, expected '1' for "1010..." pattern
                if (currentChar == '0') {
                    changesForStart1++; // Mismatch, needs a change
                }
            } else { // Odd index, expected '0' for "1010..." pattern
                if (currentChar == '1') {
                    changesForStart1++; // Mismatch, needs a change
                }
            }
        }

        // The minimum number of operations is the smaller of the two counts.
        return Math.min(changesForStart0, changesForStart1);
    }
}
```

File: solutions/2026-03-05/javascript/2026-03-05-minimum-changes-to-make-alternating-binary-string.js
// Problem: Minimum Changes To Make Alternating Binary String
// Link: https://leetcode.com/problems/minimum-changes-to-make-alternating-binary-string/
//
// Approach:
// An alternating binary string can only have two forms:
// 1. Starts with '0': "010101..."
// 2. Starts with '1': "101010..."
//
// We can iterate through the input string `s` and count the number of characters that
// do not match the expected character for each of these two alternating patterns.
//
// For the pattern starting with '0' (e.g., "010101..."):
// At even indices (0, 2, 4, ...), the character should be '0'.
// At odd indices (1, 3, 5, ...), the character should be '1'.
// We count how many characters in `s` deviate from this pattern. Let this be `changesForZeroStart`.
//
// For the pattern starting with '1' (e.g., "101010..."):
// At even indices (0, 2, 4, ...), the character should be '1'.
// At odd indices (1, 3, 5, ...), the character should be '0'.
// We count how many characters in `s` deviate from this pattern. Let this be `changesForOneStart`.
//
// The minimum number of operations will be the minimum of `changesForZeroStart` and `changesForOneStart`.
//
// Time Complexity: O(n), where n is the length of the string `s`. We iterate through the string once.
// Space Complexity: O(1). We only use a few variables to store counts.

/**
 * @param {string} s
 * @return {number}
 */
const minChanges = (s) => {
    // Initialize counts for deviations from two possible alternating patterns.
    // changesForZeroStart: counts deviations if the target string starts with '0' (e.g., "0101...")
    // changesForOneStart: counts deviations if the target string starts with '1' (e.g., "1010...")
    let changesForZeroStart = 0;
    let changesForOneStart = 0;

    // Iterate through the input string `s` using its index `i`.
    for (let i = 0; i < s.length; i++) {
        // Check for the pattern starting with '0'.
        // If the index `i` is even, the expected character is '0'.
        // If the index `i` is odd, the expected character is '1'.
        if (i % 2 === 0) {
            // At an even index, if the character is '1', it deviates from the '0' start pattern.
            if (s[i] === '1') {
                changesForZeroStart++;
            }
        } else {
            // At an odd index, if the character is '0', it deviates from the '0' start pattern.
            if (s[i] === '0') {
                changesForZeroStart++;
            }
        }

        // Check for the pattern starting with '1'.
        // If the index `i` is even, the expected character is '1'.
        // If the index `i` is odd, the expected character is '0'.
        if (i % 2 === 0) {
            // At an even index, if the character is '0', it deviates from the '1' start pattern.
            if (s[i] === '0') {
                changesForOneStart++;
            }
        } else {
            // At an odd index, if the character is '1', it deviates from the '1' start pattern.
            if (s[i] === '1') {
                changesForOneStart++;
            }
        }
    }

    // The minimum number of operations is the smaller of the two deviation counts.
    return Math.min(changesForZeroStart, changesForOneStart);
};
```